### PR TITLE
[CloudSync] Adding the API to be exposed to thirdparty for Data Publish

### PR DESCRIPTION
**Pull request by nitu-s-gupta**
Signed-off-by: Nitu Gupta <nitu.gupta@samsung.com>

# Description
provide mechanism to expose API that can be used by third party apps to publish data to MQTT broker running on AWS endpoint

Fixes # (#382 )

## Type of change
- [X] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?
1.MQTT mosquitto broker is configured to be running in the AWS endpoint.
2. The edge orchestration is build and run using following command with option CLOUD_SYNC set to true
```
docker run -it -d --privileged --network="host" --name edge-orchestration -e CLOUD_SYNC=true -v /var/edge-orchestration/:/var/edge-orchestration/:rw -v /var/run/docker.sock:/var/run/docker.sock:rw -v /proc/:/process/:ro  lfedge/edge-home-orchestration-go:latest
```
3. The logs are checked using the command
```
docker logs -f edge-orchestration
```
4. From the another terminal/post make a curl command as follows to publish data using home edge to the broker running on AWS endpoint
```
curl --location --request POST 'http://192.168.1.107:56001/api/v1/orchestration/cloudsyncmgr' \
> --header 'Content-Type: text/plain' \
> --data-raw '{
>     "appid": "TV1",
>     "payload": "{Another data from TV1}",
>     "topic": "home/livingroom",
>     "url" : "ec2-54-144-110-26.compute-1.amazonaws.com"
> }'
```
5. You should obtain the following message for successful publish
```
{"Message":"Data published successfully to Cloud","RemoteTargetInfo":"","ServiceName":""}
```

**Test Configuration**:
* OS type & version: (Ubuntu 20.04)
* Hardware: (x86-64)
* Edge Orchestration Release: (v1.1.0)

# Checklist:

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: internal/restinterface/externalhandler/externalhandler.go [hunk 1]
```diff
@@ -351,6 +359,101 @@ SEND_RESP:
 	h.helper.Response(w, respEncryptBytes, http.StatusOK)
 }
 
+// APIV1RequestCloudSyncmgrPost handles cloudsync publish request from service application
+func (h *Handler) APIV1RequestCloudSyncmgrPost(w http.ResponseWriter, r *http.Request) {
```
Comment (t25kim) [inline internal/restinterface/externalhandler/externalhandler.go:363]: If I call this 10 times with the same host, will 10 mqtt client go routines be created?

**Review by t25kim — COMMENTED** [review]
Thanks for the contribution. 
Please take a look at my comments.

File: internal/restinterface/externalhandler/externalhandler.go [hunk 2]
```diff
@@ -351,6 +359,101 @@ SEND_RESP:
...
+		goto SEND_RESP
+	}
+	url, ok = appCommand["url"].(string)
+	if !ok {
+		responseMsg = orchestrationapi.InvalidParameter
+		goto SEND_RESP
+	}
+	cloudsyncmgr.GetInstance().StartCloudSync(url)
+	for mqttmgr.GetClient() == nil {
+		//Wait till the client is set
```
[the remark this answers is not in the document]
Comment (nitu-s-gupta) [reply internal/restinterface/externalhandler/externalhandler.go:437]: If using sleep what time value would be used. Its not always same in settting client

File: internal/restinterface/externalhandler/externalhandler.go (hunk 1 above)
Comment (nitu-s-gupta) [reply internal/restinterface/externalhandler/externalhandler.go:363]: currently it is not storing the client data. Also one more thing when same host is sending the conenctn request the previous connection is closed and new is opened.

Comment (suresh-lc) [reply internal/restinterface/externalhandler/externalhandler.go:363]: Wanted to make the base code with single connection with the client. Going forward we think of scenarios where multiple service applications can call the send API to send data to different broker URL's. So in that case maintaining separate mqtt client instance for each service application would be better(and only one instance for each host). Use of db to maintain the same or so could be option. We plan to do this once the base code is merged as part of next steps.

**Review by tdrozdovsky — COMMENTED** [review]
LGTM

**Review by tdrozdovsky — APPROVED** [review]

**Review by t25kim — APPROVED** [review]
LGTM Thanks!

**Review by suresh-lc — APPROVED** [review]
LGTM

**Review by MoonkiHong — APPROVED** [review]
+1

**Pull request merged**